python/cudf/cudf/core/internals/where.py

```diff
@@ -0,0 +1,375 @@
+        Whether to perform the operation in place on the data.
+
+    Returns
+    -------
+    Same type as caller
+
+    Examples
+    --------
+    >>> import cudf
+    >>> df = cudf.DataFrame({"A":[1, 4, 5], "B":[3, 5, 8]})
+    >>> df.where(df % 2 == 0, [-1, -1])
+        A  B
+    0 -1 -1
+    1  4 -1
+    2 -1  8
+
+    >>> ser = cudf.Series([4, 3, 2, 1, 0])
+    >>> ser.where(ser > 2, 10)
+    0     4
+    1     3
+    2    10
+    3    10
+    4    10
+    dtype: int64
+    >>> ser.where(ser > 2)
+    0       4
+    1       3
+    2    <NA>
+    3    <NA>
+    4    <NA>
+    dtype: int64
+    """
+
+    if isinstance(frame, cudf.DataFrame):
+        if hasattr(cond, "__cuda_array_interface__"):
+            cond = cudf.DataFrame(
+                cond, columns=frame._data.names, index=frame.index
+            )
+        elif not isinstance(cond, cudf.DataFrame):
+            cond = frame.from_pandas(pd.DataFrame(cond))
+
+        common_cols = set(frame._data.names).intersection(
+            set(cond._data.names)
+        )
+        if len(common_cols) > 0:
+            # If `frame` and `cond` are having unequal index,
+            # then re-index `cond`.
+            if not frame.index.equals(cond.index):
+                cond = cond.reindex(frame.index)
+        else:
+            if cond.shape != frame.shape:
+                raise ValueError(
+                    """Array conditional must be same shape as self"""
+                )
+            # Setting `frame` column names to `cond`
+            # as `cond` has no column names.
+            cond.columns = frame.columns
+
+        (source_df, others,) = _normalize_columns_and_scalars_type(
+            frame, other
+        )
+        if isinstance(other, cudf.core.frame.Frame):
+            others = others._data.columns
+
+        out_df = cudf.DataFrame(index=frame.index)
+        if len(frame._columns) != len(others):
+            raise ValueError(
+                """Replacement list length or number of dataframe columns
+                should be equal to Number of columns of dataframe"""
+            )
+        for i, column_name in enumerate(frame._data.names):
+            input_col = source_df._data[column_name]
+            other_column = others[i]
+            if column_name in cond._data:
+                if isinstance(input_col, cudf.core.column.CategoricalColumn):
+                    if cudf.utils.dtypes.is_scalar(other_column):
+                        try:
+                            other_column = input_col._encode(other_column)
+                        except ValueError:
+                            # When other is not present in categories,
+                            # fill with Null.
+                            other_column = None
+                        other_column = cudf.Scalar(
+                            other_column, dtype=input_col.codes.dtype
+                        )
+                    elif hasattr(other_column, "codes"):
```

Comment:
Any reason we're using a `hasattr` check here instead of `isinstance(other_column, CategoricalColumn)`?